pandas/io/xml.py

```diff
@@ -0,0 +1,892 @@
+        stylesheet,
+    ):
+        self.io = io
+        self.xpath = xpath
+        self.namespaces = namespaces
+        self.elems_only = elems_only
+        self.attrs_only = attrs_only
+        self.names = names
+        self.encoding = encoding
+        self.stylesheet = stylesheet
+        self.is_style = None
+
+    def parse_data(self) -> List[Dict[str, Optional[str]]]:
+        """
+        Parse xml data.
+
+        This method will call the other internal methods to
+        validate xpath, names, parse and return specific nodes.
+        """
+
+        raise AbstractMethodError(self)
+
+    def _parse_nodes(self) -> List[Dict[str, Optional[str]]]:
+        """
+        Parse xml nodes.
+
+        This method will parse the children and attributes of elements
+        in xpath, conditionally for only elements, only attributes
+        or both while optionally renaming node names.
+
+        Raises
+        ------
+        ValueError
+            * If only elements and only attributes are specified.
+
+        Notes
+        -----
+        Namespace URIs will be removed from return node values.Also,
+        elements with missing children or attributes compared to siblings
+        will have optional keys filled withi None values.
+        """
+
+        raise AbstractMethodError(self)
+
+    def _validate_path(self) -> None:
+        """
+        Validate xpath.
+
+        This method checks for syntax, evaluation, or empty nodes return.
+
+        Raises
+        ------
+        SyntaxError
+            * If xpah is not supported or issues with namespaces.
+
+        ValueError
+            * If xpah does not return any nodes.
+        """
+
+        raise AbstractMethodError(self)
+
+    def _validate_names(self) -> None:
+        """
+        Validate names.
+
+        This method will check if names is a list-like and aligns
+        with length of parse nodes.
+
+        Raises
+        ------
+        ValueError
+            * If value is not a list and less then length of nodes.
+        """
+        raise AbstractMethodError(self)
+
+    def _convert_io(self, xml_data) -> Union[str, bytes, None]:
+        """
+        Convert io object to string.
+
+        This method will convert io object into a string or keep
+        as string, depending on object type.
+        """
+
+        obj: Union[bytes, str, None] = None
+
+        if isinstance(xml_data, str):
+            obj = xml_data
+
+        elif isinstance(xml_data, bytes):
+            obj = xml_data.decode(self.encoding)
+
+        elif isinstance(xml_data, io.StringIO):
+            obj = xml_data.getvalue()
+
+        elif isinstance(xml_data, io.BytesIO):
+            obj = xml_data.getvalue().decode(self.encoding)
+
+        elif isinstance(xml_data, io.TextIOWrapper):
+            obj = xml_data.read()
+
+        elif isinstance(xml_data, io.BufferedReader):
+            obj = xml_data.read().decode(self.encoding)
+        else:
+            obj = None
```

Comment:
`os.PathLike` does not seem to be handled. Are buffers such as `fsspec`/`s3` handled by this?

Reply:
How about your suggestion above with `if isinstance(stringify_path(xml_data), str)` for first condition which passes tests?

Reply:
sorry, you are right - `PathLike` is handled by `stringify_path`. I didn't see that call :) I'm not 100% sure that the current code deals with all file buffers.